### DISCUSSION TITLE
Configuring Directus with external Postgres Database Cluster (Digitalocean)

### DISCUSSION THREAD

**Discussion by coignite-user** [directus/directus#6832] [Archive]
Does anyone have experience or can help to connect Directus to an external database.

I installed it using [One-Click installation](https://marketplace.digitalocean.com/apps/directus?action=deploy&refcode=4c0b6062c16e) and configured everything else, but not able to connect to the external database. [Here is a bit of documentation](https://docs.digitalocean.com/products/databases/postgresql/how-to/connect/).

The installation has Caddy reversed proxy in front.

My configuration is something like (dummy data). I allowed sources of the droplet to connect to the database. I can connect from within the server using psql. But the following configuration doesn't allow me to `directus bootstrap`. I simply get:

>Couldn't connect to the database

without any explanation.

Here is the configuration that fails
```
DB_CLIENT="pg"
DB_CONNECTION_STRING="postgresql://user:password@my-database-do-user-1111-0.b.db.ondigitalocean.com:25060/defaultdb?sslmode=require"
DB_SSL__REJECT_UNAUTHORIZED="false"
```

**Comment by Kematia** [comment]
try ` NODE_TLS_REJECT_UNAUTHORIZED="0"` instead of `DB_SSL__REJECT_UNAUTHORIZED="false"`

**Reply by rijkvanzanten** [reply]
Also, try adding the CA certificate as per https://docs.directus.io/guides/installation/digitalocean-app-platform/#_6-configure-the-environment-variables

**Reply by coignite-user** [reply]
> try ` NODE_TLS_REJECT_UNAUTHORIZED="0"` instead of `DB_SSL__REJECT_UNAUTHORIZED="false"`

Thank you, this made it work instantly! But, now I get the following message when bootstrapping the database

```
(node:1139) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
```

What does ` NODE_TLS_REJECT_UNAUTHORIZED="0"` do exactly?

Does this mean that all API requests can be unsecured? Is there a way to make it secured?

**Reply by coignite-user** [reply]
> Also, try adding the CA certificate as per https://docs.directus.io/guides/installation/digitalocean-app-platform/#_6-configure-the-environment-variables

Thank you. I added that to .env

```
NODE_TLS_REJECT_UNAUTHORIZED="0"
DB_SSL__REJECT_UNAUTHORIZED=true
DB_SSL__CA=${my-database-cluster-name.CA_CERT}
```

I bootstrapped the database and restarted, but I don't know if there is an effect with this.

I also tried to remove  `NODE_TLS_REJECT_UNAUTHORIZED=true`
```
#NODE_TLS_REJECT_UNAUTHORIZED="0"
DB_SSL__REJECT_UNAUTHORIZED=true
DB_SSL__CA=${my-database-cluster-name.CA_CERT}
```

but then bootstrap fails with

> Can't connect to the database

What does this line actually do `DB_SSL__CA=${my-database-cluster-name.CA_CERT}`? How is it treated, is it replaced and where?

**Reply by coignite-user** [reply]
There is an option to download CA certificate on the Postgres Database Cluster instance (Digitalocean). Will it make sense to feed it on the Directus server, so the requests to database a signed with it somehow?

**Reply by rijkvanzanten** [reply]
@rassvetdk Make sure to replace the `my-database-cluster-name` with the actual name of your database service in the app context

**Reply by rijkvanzanten** [reply]
> There is an option to download CA certificate on the Postgres Database Cluster instance (Digitalocean). Will it make sense to feed it on the Directus server, so the requests to database a signed with it somehow?

You don't have to do that manually. The CA CERT contents is available in that env var I shared earlier 🙂

**Reply by coignite-user** [reply]
> > There is an option to download CA certificate on the Postgres Database Cluster instance (Digitalocean). Will it make sense to feed it on the Directus server, so the requests to database a signed with it somehow?
> 
> You don't have to do that manually. The CA CERT contents is available in that env var I shared earlier 🙂

Yes, that is also what I thought, but do you agree that it should work without setting `NODE_TLS_REJECT_UNAUTHORIZED="0"`?

And, this `my-database-cluster-name` is it the one marked with **bold** belov, or something else?

postgresql://user:password@**my-database**-do-user-1111-0.b.db.ondigitalocean.com:25060/defaultdb?sslmode=require

**Reply by rijkvanzanten** [reply]
> Yes, that is also what I thought, but do you agree that it should work without setting  NODE_TLS_REJECT_UNAUTHORIZED="0"?

Yes, that's correct (and it does, that's how I use it myself too 🙂)

You only need that above flag if the CA CERT is invalid/missing.

> And, this my-database-cluster-name is it the one marked with bold belov, or something else?

It's the name of the "component" in your app platform instance, you should be able to find it under the Settings tab:

![CleanShot 2021-07-19 at 09 36 43@2x](https://user-images.githubusercontent.com/9141017/126168573-9715ccc2-0453-41cc-8d15-eb70942f7127.png)

So in this example, my "component" is named "internal-data", so the env var would be `${internal-data.CA_CERT}`

**Reply by coignite-user** [reply]
> > Yes, that is also what I thought, but do you agree that it should work without setting  NODE_TLS_REJECT_UNAUTHORIZED="0"?
> 
> Yes, that's correct (and it does, that's how I use it myself too 🙂)
> 
> You only need that above flag if the CA CERT is invalid/missing.
> 
> > And, this my-database-cluster-name is it the one marked with bold belov, or something else?
> 
> It's the name of the "component" in your app platform instance, you should be able to find it under the Settings tab:
> 
> ![CleanShot 2021-07-19 at 09 36 43@2x](https://user-images.githubusercontent.com/9141017/126168573-9715ccc2-0453-41cc-8d15-eb70942f7127.png)
> 
> So in this example, my "component" is named "internal-data", so the env var would be `${internal-data.CA_CERT}`

I am not using App platform (maybe I should) I just used Marketplace One-Click droplet with preinstalled Directus + Caddy (reversed proxy) and I am trying to change .env file to connect to the managed database privately through VPC network. In this configuration, I don't have any Components as such. Maybe I should try to make a setup using App feature for the fun of it, as making the extensions for Directus is the next step and your setup seems to be more convenient for that. And this way I will have to build the deployment pipe myself.

Thank you @rijkvanzanten! Do you care to share any good resources for the setup above? I believe it can be useful not only for me but anyone running into this kind of issue. Have a good evening, and thanks once again :)

**Reply by rijkvanzanten** [reply]
Ohhh my bad! In that case, you'll have to download that CA file manually like you suggested, and put it on the droplet so you can read it into the env var for Directus. 

I would recommend trying out App Platform for this use case, as it takes a way a lot of manual hassle you normally have to deal with with a droplet (reverse proxies, CDN, SSL, etc etc et)

**Reply by rijkvanzanten** [reply]
As for resources, I wrote a (very barebones) step by step on how to get Directus up and running on App Platform here: https://docs.directus.io/guides/installation/digitalocean-app-platform/

**Reply by coignite-user** [reply]
Thank you for sharing, now I know who did it :)

Just to finish up this thread. Do you know where I can see any details about what hides behind "Can't connect to the database"?  Are there any log files? I am just curious to find out what exactly is failing.

![CleanShot 2021-07-19 at 20 55 39@2x](https://user-images.githubusercontent.com/28866431/126212662-6d61f069-d838-48ef-9926-91a35688b66b.png)

I managed to download the CA and upload it  into `~/.postgresql/root.cert` folder. I am able to connect from the droplet using psql command with **sslmode=verify-full**, but I am not able to setup Directus's `DB_SSL__CA` variable properly to make it work. The last try was to write 

`DB_SSL__CA=~/.postgresql/root.cert`

Still the same `Can't connect to the database`, which starts being annoying without ability to dive deeper :)

**Reply by kropsi** [reply]
Hi @rijkvanzanten, I (aka @rassvetdk) am running a bit in circles here. Now I have a App platform installation with nodejs and the managed database commonent. Still I get the same `Can't connect to the database` error. I have the following app spec. Any ideas. Now that you have something of the similar setup, maybe you can pipoint the problem.

```yaml
databases:
- cluster_name: my-database-live
  db_name: defaultdb
  db_user: doadmin
  engine: PG
  name: my-database-live
  production: true
  version: "13"
name: -cms
region: fra
services:
- build_command: npx directus bootstrap
  environment_slug: node-js
  envs:
  - key: DB_SSL__REJECT_UNAUTHORIZED
    scope: RUN_AND_BUILD_TIME
    value: "true"
  - key: DB_SSL__CA
    scope: RUN_AND_BUILD_TIME
    value: ${my-database-live.CA_CERT}
  - key: DB_CONNECTION_STRING
    scope: RUN_AND_BUILD_TIME
    value: ${my-database-live.DATABASE_URL}
  github:
    branch: main
    deploy_on_push: true
    repo: rassvetdk/directus-cms
  http_port: 8080
  instance_count: 1
  instance_size_slug: basic-xs
  name: directus-cms
  routes:
  - path: /
  run_command: npx directus start
  source_dir: /
```

`Can't connect to the database` error doesn't tell anythihg, so it is difficult to understand what exactly fails, but it is something of the same.

You write [here](https://docs.directus.io/guides/installation/digitalocean-app-platform/#_5-setup-the-build-step) that `npx directus bootstrap`should be a part of the build command, but [here](https://docs.digitalocean.com/products/app-platform/how-to/use-environment-variables/#databases) it states that database variables are not available at `BUILD_TIME`. It is a bit confusing.

I hope maybe you will bring some light on a working version of App Spec for App Platform that works, since you have a working version,

**Reply by rijkvanzanten** [reply]
Can you try bootstrapping as part of the run command? `npx directus bootstrap && npx directus start` as the run command

**Reply by kropsi** [reply]
I traced that environment variables are not initialised at build time. They are only available at run time. I also contacted Digitalocean support and they told me that database_url or connection string is not available in App Platform, so it must be connected with host, user and pass. I’ve done everything and exactly what you are writing, but I still get the same error. I traced that all environment variables are as they suppose to be at the moment that I run `npx directus bootstrap && npx directus start`, bot no luck.

I can see I am not alone having this trouble on the forum, it could be so nice to get a bit more information why connection fails, besides that it fails.

I believe that it is something to do with SSL, but I also traced that it is feed into `DB_SSL__CA` variable and it is.

I am not quite sure how to approach this and who to ask about this.

**Reply by rijkvanzanten** [reply]
>  it could be so nice to get a bit more information why connection fails, besides that it fails.

Coming right up: https://github.com/directus/directus/pull/6977

**Reply by kropsi** [reply]
> > it could be so nice to get a bit more information why connection fails, besides that it fails.
> 
> Coming right up: #6977

Thank you @rijkvanzanten !

**Reply by arladmin** [reply]
> Thank you for sharing, now I know who did it :)
> 
> Just to finish up this thread. Do you know where I can see any details about what hides behind "Can't connect to the database"? Are there any log files? I am just curious to find out what exactly is failing.
> 
> ![CleanShot 2021-07-19 at 20 55 39@2x](https://user-images.githubusercontent.com/28866431/126212662-6d61f069-d838-48ef-9926-91a35688b66b.png)
> 
> I managed to download the CA and upload it into `~/.postgresql/root.cert` folder. I am able to connect from the droplet using psql command with **sslmode=verify-full**, but I am not able to setup Directus's `DB_SSL__CA` variable properly to make it work. The last try was to write
> 
> `DB_SSL__CA=~/.postgresql/root.cert`
> 
> Still the same `Can't connect to the database`, which starts being annoying without ability to dive deeper :)

@rassvetdk Did you manage to get it going?
I'm in the same situation currently. Unable to connect my docker setup with DO managed postgres due to the certificate issue.

**Reply by kropsi** [reply]
@arladmin yes, please read all the comments in this thread. After #6977 the tool tells you more about what is behind the error, you probably need to upgrade directus to the latest version.

**Reply by arladmin** [reply]
> @arladmin yes, please read all the comments in this thread. After #6977 the tool tells you more about what is behind the error, you probably need to upgrade directus to the latest version.

Thanks, i'm already on the latest version though, v9.0.0-rc.87, and it didn't work.
The reason being, i was using the DB connection string to configure my database (not something mentioned in the comments above, or the logs :wink: ).

My connection string:
```
DB_SSL__CA="./ca-certificate.crt"
DB_CONNECTION_STRING=postgresql://user:password@db-host:port/db-name?sslmode=require
```
It did finally work when i used the individual ENV parameters instead.
```
DB_SSL__CA="./ca-certificate.crt"
DB_SSL__REJECT_UNAUTHORIZED=false
DB_CLIENT=pg
DB_HOST=db-host
DB_PORT=port
DB_DATABASE=db-name
DB_USER=user
DB_PASSWORD=password
DB_SSLMODE=require
```

**Reply by kropsi** [reply]
@arladmin connection string is currently not supported by App Platform (DO)

**Reply by arladmin** [reply]
> > Thank you for sharing, now I know who did it :)
> > Just to finish up this thread. Do you know where I can see any details about what hides behind "Can't connect to the database"? Are there any log files? I am just curious to find out what exactly is failing.
> > ![CleanShot 2021-07-19 at 20 55 39@2x](https://user-images.githubusercontent.com/28866431/126212662-6d61f069-d838-48ef-9926-91a35688b66b.png)
> > I managed to download the CA and upload it into `~/.postgresql/root.cert` folder. I am able to connect from the droplet using psql command with **sslmode=verify-full**, but I am not able to setup Directus's `DB_SSL__CA` variable properly to make it work. The last try was to write
> > `DB_SSL__CA=~/.postgresql/root.cert`
> > Still the same `Can't connect to the database`, which starts being annoying without ability to dive deeper :)
> 
> @rassvetdk Did you manage to get it going?
> I'm in the same situation currently. Unable to connect my docker setup with DO managed postgres due to the certificate issue.

@kropsi 
Like i mentioned here, my setup is Docker (standalone) based, not App Platform.

**Comment by kropsi** [comment]
I managed to configure my setup with App Platform, but not with Digital Ocean one-click-installation. When configuring one click-installation there is some self-signed certificate that got in my way when trying to get it to work. #6977 helped identify that issue. I believe that there is definitely a way to make it to work, but it requires a different system administration skill set than mine.

App platform works perfectly for that matter, as it does a lot of configuration for you and takes care of certificates. I experienced a problem with App platform as well, and after a lot of research, it appeared to be something very simple. When App platform re-builds it assigns a new IP for your app each and every time. Therefore you cannot add it to trusted sources of your managed database and expect it to work after re-build. It is obviously a known issue (https://ideas.digitalocean.com/ideas/APPX-I-73) which I was not aware of. Securing the connection between App platform and the Cloud database is on the way.

**Reply by luanmm** [reply]
@kropsi could you share with us how you managed to setup Directus in App Platform? Following the steps in the docs I couldn't solve my problems in the environment.

I'm deploying Directus from a repository (with some customizations), but couldn't manage to solve the error when connecting to PostgreSQL. The database was created in the Digital Ocean's environment, but "outside" App Platform (is a standard managed DB instance, for the production environment).

The only way I could connect to it from App Platform was setting NODE_TLS_REJECT_UNAUTHORIZED environment variable. Every other configuration that I used was getting errors. Couldn't manage to solve using DB_SSL__CA, for instance, so, for now, I'm with this "workaround", with security issues that I'm concerned.

**Reply by kropsi** [reply]
@luanmm we have the same setup I believe. I experienced that when NODE_TLS_REJECT_UNAUTHORIZED was true I had no connection and when false I had a connection. I saw that there was a connection to the database and then it was instantly dropped.

Now if you in your cluster database have Trusted sources set to any specific IP it automatically excludes all others who are trying to connect to it. I currently removed all trusted sources from my database (which actually expose the database to the public) I did it knowing that my database is not critical and that DO is working on solving this with high pressure from the community. I believe you need to check your trusted sources and try to reconnect. At some point, I had to completely wipe my App Platform to make it work, but because I put all the critical Env variables into the App Platform App spec, I can wipe and create it anytime now and it works each time.

If you have critical data, you might need to investigate a proxy solution for the database.

**Reply by luanmm** [reply]
And the only related environment variables you left in your setup was the two from Directus docs?

```env
DB_SSL__REJECT_UNAUTHORIZED=true
DB_SSL__CA=${my-database-component-name.CA_CERT}
```

Like this?
If this is the case, maybe my configuration didn't work because between tests I never wiped the app to start from scratch. Maybe there is some cache in App Platform or something like that.

**Reply by kropsi** [reply]
@luanmm I also have all the database connection variables as well
```yaml
envs:
- key: DB_HOST
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.HOSTNAME}
- key: DB_SSL__REJECT_UNAUTHARIZED
  scope: RUN_AND_BUILD_TIME
  value: "true"
- key: DB_PORT
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.PORT}
- key: DB_DATABASE
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.DATABASE}
- key: DB_USER
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.USERNAME}
- key: DB_PASSWORD
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.PASSWORD}
- key: DB_SSL__CA
  scope: RUN_AND_BUILD_TIME
  value: ${database-component.CA_CERT}
```
It might help to wipe and start over. I also experienced that as if there were some caching in the configuration, but I am not sure.

**Reply by kropsi** [reply]
**BTW!** the connection string is not supported in App Platform (DO) at the moment
```yaml
envs:
  - key: DB_SSL__REJECT_UNAUTHORIZED
    scope: RUN_AND_BUILD_TIME
    value: "true"
  - key: DB_SSL__CA
    scope: RUN_AND_BUILD_TIME
    value: ${database-component.CA_CERT}
  - key: DB_CONNECTION_STRING
    scope: RUN_AND_BUILD_TIME
    value: ${database-component.DATABASE_URL}
```

**Reply by luanmm** [reply]
Thank you very much for the detailed comments, @kropsi! I'm sure it will help a lot (me and others) until App Platform can be upgraded by DO's team to simplify this things.

I will try to set again the environment variables using your sample as a start. Thanks again! :smiley:
